PR_Epic3_43 data layer for favorite

danil377377 opened the pull request:
Добавил всё по этим критериям
![image](https://github.com/user-attachments/assets/6edcf919-a860-4088-8dc6-04651067242a)
, кроме последнего(добавление во viewModel), так как этим собираюсь заняться во время работы с Presentation в другой задаче

danil377377 committed:
добавление метода удаления вакансии из БД

danil377377 committed:
добавление класса вакакнсии и создание интерфейсов репозитория и интерактора

danil377377 committed:
реализовал интерактор и репозиторий. Добавил их инициализацию в DI

danil377377 committed:
поменял расположение директорий и добавил функцию обновления вкансии

sobolevkir commented on app/src/main/java/ru/practicum/android/diploma/common/domain/model/Vacancy.kt:
```diff
@@ -0,0 +1,15 @@
+    val experience: String?,
+    val schedule: String?,
+    val description: String?,
+    val keySkills: String?,
```
вот здесь и везде где есть списки лучше сделать List<String>

sobolevkir replied:
Кроме этого некоторые поля ошибочно указаны как Nullable, хотя сервер их возвращает по required.
Имя этого класса будет VacancyDetails.
Также не вижу в этой модели данных поля address и contacts. Более подробно про них написал в пачке, но предлагаю сделать вот так:
```

data class VacancyDetails(
    val id: Long,
    val name: String,
    val salary: Salary?,
    val areaName: String,
    val employerName: String,
    val employerLogoUrl240: String?,
    val contacts: Contacts?,
    val experience: String?,
    val scheduleName: String,
    val description: String,
    val keySkills: List<String>,
    val address: Address?,
    val isFavorite: Boolean = false,
)
data class Salary(
    val currency: String?,
    val from: Double?,
    val to: Double?
)
data class Contacts(
    val name: String?,
    val email: String?,
    val phones: List<String>?
)
data class Address(
    val city: String?,
    val street: String?,
    val building: String?
)
```

sobolevkir commented on app/src/main/java/ru/practicum/android/diploma/favorites/data/converter/FavoriteVacancyDbConverter.kt:
через этот конвертер и gson объект из коина мы можем сконвертировать String из БД в List<String> и обратно!

sobolevkir commented on app/src/main/java/ru/practicum/android/diploma/favorites/data/dao/FavoriteVacancyDao.kt:
не нужно ли в методах remove, insert и update сделать возвращение Int/Long (какой именно тип возвращают эти методы можно почитать в gpt или документации к Room. Вдруг где-то нужно будет сделать проверку и убедиться в том, что у нас операция прошла успешно?

sobolevkir commented on app/src/main/java/ru/practicum/android/diploma/favorites/data/impl/FavoritesRepositoryImpl.kt:
соответственно репозиторий может тоже возвращать значение для проверки успешности операций. но я не уверен насчёт этого, нужно посмотреть логику приложения, есть ли где-то потребность в таких проверках?)

danil377377 committed:
немного модифицировал класс вакансии.
p.s.Решил что не нужно добавлять возвращаемые типы, так как проверить на всё ли нормально взялось из БД можно с помощью исключений. То есть при пустом листе вакансий из БД мы выведем соответсвующий плейсхолдер, а при эксепшене другой

danil377377 committed:
пришось изменить класс коныертера, из за правил рума

danil377377 committed:
исправление детект

Yuliya-Sycheva commented on app/src/main/java/ru/practicum/android/diploma/favorites/data/converter/FavoriteVacancyDbConverter.kt:
```diff
@@ -0,0 +1,39 @@
+        gson.fromJson(favoriteVacancyEntity.keySkills, object : TypeToken<List<String>>() {}.type),
+        isFavorite = true
+    )
+}
```
Можно использовать объект с extension-функциями, тогда не придется создавать в Koin конвертер и передавать в конструкторе в FavoritesRepositoryImpl. И вызывать удобнее.
Можно сделать по аналогии с:

```
object FavoriteTrackDbConvertor {
    fun Track.toFavoriteTrackEntity(): FavoriteTrackEntity {
        return FavoriteTrackEntity(
            trackId,
            trackName,
            artistName,
            trackTimeMillis,
            artworkUrl100,
            collectionName,
            releaseDate,
            primaryGenreName,
            country,
            previewUrl
        )
    }
    fun FavoriteTrackEntity.toTrack(): Track {
        return Track(
            trackId,
            trackName,
            artistName,
            trackTime,
            artworkUrl100,
            collectionName,
            releaseDate,
            primaryGenreName,
            country,
            previewUrl,
            isFavorite = true 
        )
    }
}
```

sobolevkir replied:
я бы не стал так делать. синглтон- это антипаттерн. удобнее коин-модули смотреть будет, где есть 

> Можно использовать объект с extension-функциями, тогда не придется создавать в Koin конвертер и передавать в конструкторе в FavoritesRepositoryImpl. И вызывать удобнее. Можно сделать по аналогии с:
> 
> ```
> object FavoriteTrackDbConvertor {
>     fun Track.toFavoriteTrackEntity(): FavoriteTrackEntity {
>         return FavoriteTrackEntity(
>             trackId,
>             trackName,
>             artistName,
>             trackTimeMillis,
>             artworkUrl100,
>             collectionName,
>             releaseDate,
>             primaryGenreName,
>             country,
>             previewUrl
>         )
>     }
>     fun FavoriteTrackEntity.toTrack(): Track {
>         return Track(
>             trackId,
>             trackName,
>             artistName,
>             trackTime,
>             artworkUrl100,
>             collectionName,
>             releaseDate,
>             primaryGenreName,
>             country,
>             previewUrl,
>             isFavorite = true 
>         )
>     }
> }
> ```

Я бы не стал object'ы плодить. синглтон - это вроде как антипаттерн... А в коин-модулях удобно посмотреть и отследить все зависимости.

Yuliya-Sycheva commented on app/src/main/java/ru/practicum/android/diploma/favorites/data/impl/FavoritesRepositoryImpl.kt:
```diff
@@ -0,0 +1,31 @@
+
+class FavoritesRepositoryImpl(
+    private val appDatabase: AppDatabase,
+    private val vacancyDbConverter: FavoriteVacancyDbConverter
```
Если сделать extension-функции, то можно строку c vacancyDbConverter убрать. 
А вызывать методы в FavoritesRepositoryImpl по аналогии с: 
```
override suspend fun likeTrack(track: Track) {
        appDatabase.getFavoriteTrackDao().insertFavoriteTrack(track.toFavoriteTrackEntity())
    }
```

danil377377 committed:
удаление для более лёгкого слияния

danil377377 committed:
Merge branch 'develop' into Epic3_43_Data-layer-for-Favorite

danil377377 committed:
Merge branch 'develop' into Epic3_43_Data-layer-for-Favorite

danil377377 committed:
обновил классы

danil377377 committed:
обновил репозитории и интеракторы

danil377377 committed:
поправки, чтобы собралось

danil377377 committed:
Merge branch 'develop' into Epic3_43_Data-layer-for-Favorite
